apollo-compiler/src/main/kotlin/com/apollographql/apollo3/compiler/codegen/CodegenLayout.kt

```diff
@@ -17,13 +20,62 @@ import com.apollographql.apollo3.compiler.singularize
  *
  * Inputs should always be GraphQL identifiers and outputs are valid Kotlin/Java identifiers.
  */
-abstract class CodegenLayout(
+internal abstract class CodegenLayout(
+    private val ir: Ir,
     private val packageNameGenerator: PackageNameGenerator,
     private val schemaPackageName: String,
     private val useSemanticNaming: Boolean,
     private val useSchemaPackageNameForFragments: Boolean,
-    private val nameToClassName: (String) -> String,
 ) {
+  private val schemaTypeToClassName: Map<IrSchemaType, String> = computeClassNames()
+
+  private fun computeClassNames(): Map<IrSchemaType, String> {
+    val schemaTypeToClassName = mutableMapOf<IrSchemaType, String>()
+    val allTypes: List<IrSchemaType> = ir.customScalars + ir.objects + ir.enums + ir.interfaces + ir.inputObjects + ir.unions
+    val usedNames = mutableSetOf<String>()
+    for (type in allTypes) {
+      val targetName = type.targetName
+      val className = if (targetName != null) {
+        // If a targetName was specified, honor it verbatim
+        targetName
+      } else {
+        val uniqueName = uniqueName(type.name, usedNames)
+        // Capitalize all types except enums
```

Comment:
This smells. I can't really remember why such a decision was made 🤔 . But it feels asymmetrical...

My guess is that this predates the "case insensitive filesystem" issues and the assumption was to allow different enums like `Direction` and `direction` but the same is true about other schema types too

Reply:
Certainly something to change in 4.x